packages/radio-group/src/vaadin-radio-group.js

```diff
@@ -73,510 +89,378 @@ class RadioGroupElement extends ThemableMixin(DirMixin(PolymerElement)) {
-        e.preventDefault();
-        this._selectIncButton(!horizontalRTL, checkedRadioButton);
-      }
-    });
-
-    this.addEventListener('focusin', () => this._setFocused(this._containsFocus()));
-
-    this.addEventListener('focusout', () => {
-      this.validate();
-      this._setFocused(false);
-    });
+  /**
+   * @type {!boolean}
+   * @private
+   */
+  get isHorizontalRTL() {
+    return this.getAttribute('dir') === 'rtl' && this.theme !== 'vertical';
   }
 
   /**
-   * @param {boolean} next
-   * @param {!RadioButton} checkedRadioButton
+   * @type {!string}
+   * @override
    * @protected
    */
-  _selectIncButton(next, checkedRadioButton) {
-    if (next) {
-      this._selectNextButton(checkedRadioButton);
-    } else {
-      this._selectPreviousButton(checkedRadioButton);
-    }
+  get _ariaAttr() {
+    return 'aria-labelledby';
   }
 
   /**
-   * @param {!RadioButton} element
-   * @param {boolean=} setFocusRing
+   * @param {KeyboardEvent} event
+   * @override
    * @protected
    */
-  _selectButton(element, setFocusRing) {
-    if (this._containsFocus()) {
-      element.focus();
-      if (setFocusRing) {
-        element.setAttribute('focus-ring', '');
-      }
+  _onKeyDown(event) {
+    super._onKeyDown(event);
+
+    const radioButton = event.composedPath().find((element) => {
+      return element instanceof RadioButton;
+    });
+
+    if (['ArrowLeft', 'ArrowUp'].includes(event.key)) {
+      event.preventDefault();
+      this.__selectNextRadioButton(radioButton);
+    }
+
+    if (['ArrowRight', 'ArrowDown'].includes(event.key)) {
+      event.preventDefault();
+      this.__selectPrevRadioButton(radioButton);
     }
-    this._changeSelectedButton(element, setFocusRing);
   }
 
   /**
-   * @return {boolean}
-   * @protected
+   * @param {!number} index
+   * @private
    */
-  _containsFocus() {
-    const activeElement = this.getRootNode().activeElement;
-    return this.contains(activeElement);
+  __selectNextRadioButton(radioButton) {
+    const index = this.__radioButtons.indexOf(radioButton);
+
+    this.__selectIncRadioButton(index, this.isHorizontalRTL ? 1 : -1);
   }
 
   /**
-   * @return {boolean}
-   * @protected
+   * @param {!number} index
+   * @private
    */
-  _hasEnabledButtons() {
-    return !this._radioButtons.every((button) => button.disabled);
+  __selectPrevRadioButton(radioButton) {
+    const index = this.__radioButtons.indexOf(radioButton);
+
+    this.__selectIncRadioButton(index, this.isHorizontalRTL ? -1 : 1);
   }
 
   /**
-   * @param {!RadioButton} element
-   * @protected
+   * @param {!number} index
+   * @param {!number} step
+   * @private
    */
-  _selectNextButton(element) {
-    if (!this._hasEnabledButtons()) {
-      return;
-    }
-
-    var nextButton = element.nextElementSibling || this.firstElementChild;
+  __selectIncRadioButton(index, step) {
+    const newIndex = (this.__radioButtons.length + index + step) % this.__radioButtons.length;
```

Comment:
The previous approach with `nextElementSibling`/ `prevElementSibling` was based on an assumption that the group contain only radio buttons as children. Now, there are also the slotted helper, label, and error message, so I had to come up with a different solution.